# Examen 1: Regresión Lineal (Ejercicio 5)

##### Taller de Modelado de Datos
##### Cristopher Corona Velasco

In [1]:
# visualización basica de los datos

import pandas as pd

PATH = "/Users/cristophercoronavelasco/Desktop/Ciencia de Datos/Taller de Modelado de Datos/data/rendimiento_atletas.csv"
df = pd.read_csv(PATH)

print(df.shape)
display(df.head(10))
df.info()
display(df.describe(include='number').T)
display(df.select_dtypes(include='object').nunique())
display(df.isna().sum())
if "sexo" in df.columns:
    display(df["sexo"].value_counts(dropna=False))

(350, 10)


edad sexo  horas_entrenamiento  vo2max  porcentaje_grasa  consumo_proteina  \
0    24    F                  3.6    49.7              15.2              1.70   
1    37    M                 10.7    64.3              12.6              1.55   
2    32    F                  7.6    43.3              22.2              1.84   
3    28    M                 12.7    61.1              21.0              1.39   
4    25    F                 11.7    58.1              17.1              1.59   
5    38    F                  5.8    43.4              20.7              1.99   
6    24    M                  6.0    41.4              11.3              1.32   
7    36    M                  5.1    59.8               8.3              1.00   
8    28    F                  4.0    58.5              19.6              1.71   
9    28    M                  6.9    55.3              14.5              1.52   

   suplementos  competencias_anuales  lesiones_previas  rendimiento  
0            1                     3                 0        47.66  
1            0                     9                 1        44.82  
2            0                     6                 3        50.47  
3            1                     6                 0        43.15  
4            0                     9                 0        40.23  
5            0                    11                 2        49.90  
6            1                     4                 3        47.06  
7            1                     5                 1        44.76  
8            0                     7                 0        45.26  
9            0                     7                 1        44.73

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   edad                  350 non-null    int64  
 1   sexo                  350 non-null    object 
 2   horas_entrenamiento   350 non-null    float64
 3   vo2max                350 non-null    float64
 4   porcentaje_grasa      350 non-null    float64
 5   consumo_proteina      350 non-null    float64
 6   suplementos           350 non-null    int64  
 7   competencias_anuales  350 non-null    int64  
 8   lesiones_previas      350 non-null    int64  
 9   rendimiento           350 non-null    float64
dtypes: float64(5), int64(4), object(1)
memory usage: 27.5+ KB


count       mean       std   min      25%    50%  \
edad                  350.0  27.948571  6.674820  18.0  22.0000  27.50   
horas_entrenamiento   350.0   8.022857  2.637484   2.0   6.2000   8.10   
vo2max                350.0  56.149429  7.678440  33.4  51.2000  55.80   
porcentaje_grasa      350.0  15.002571  4.232920   4.6  11.8500  14.90   
consumo_proteina      350.0   1.599543  0.270699   1.0   1.4200   1.61   
suplementos           350.0   0.388571  0.488123   0.0   0.0000   0.00   
competencias_anuales  350.0   6.071429  2.615369   0.0   4.0000   6.00   
lesiones_previas      350.0   1.122857  1.075538   0.0   0.0000   1.00   
rendimiento           350.0  44.939000  2.859681  38.1  43.0725  44.78   

                          75%    max  
edad                  34.0000  39.00  
horas_entrenamiento    9.9000  15.00  
vo2max                61.4000  78.40  
porcentaje_grasa      18.1000  25.40  
consumo_proteina       1.7875   2.26  
suplementos            1.0000   1.00  
competencias_anuales   8.0000  14.00  
lesiones_previas       2.0000   5.00  
rendimiento           46.9100  55.33

sexo    2
dtype: int64

edad                    0
sexo                    0
horas_entrenamiento     0
vo2max                  0
porcentaje_grasa        0
consumo_proteina        0
suplementos             0
competencias_anuales    0
lesiones_previas        0
rendimiento             0
dtype: int64

sexo
F    175
M    175
Name: count, dtype: int64

## Sección Conceptual

### Pregunta 1: Explica brevemente la diferencia entre overfitting y underfitting y describe un ejemplo de cada uno en el contexto del dataset de atletas.

- **Overfitting:** EL modelo aprende ruidos muy específicos del entrenamniento y nos daría un error muy bajo en entrenamiento y alto en testing. Un ejemplo sereía construir un modelo con muchas combinaciones ente "horas_entrenamiento", "vo2max", "porcentaje_grasa", etc. Funcionaria bien en el training pero no al evaluar.
- **Underfitting:** Es un modelo demasiado simple que no captaría la relación real , nos daríamos cuienta por errores muy grandes en el train y el test. Un ejemplo con este dataset es que si usaramos solo horas_entrenamiento para predecir el "rendimiento" e ignoraramos otras variables indispensables, no se lograría explicar bien el tiempo en 10km con la recta.

### Pregunta 2:  ¿Qué tipo de problemas podría presentar una regresión lineal simple con este dataset? Menciona dos posibles violaciones de sus supuestos y cómo podrías detectarlas.

1. **No linealidad** : La relación entre rendimiento y una sola X puede no ser una recta. Lo mas normal es que tengamos rendimientos irregulares, que disminuyen o se estabilizan a lo largo del camino. Nos daríamos cuenta con una gráfica de dispersión con una recta superpuesta, una curva significaria no linealidad.
2. **Heterocedasticidad** : El error puede ser pequeó en deportistas promedio y aumentar en atletas entrenados, lo detectariamos con un embudo en los residuos si los graficamos contra la predicción.

### Pregunta 3: Explica en tus palabras cómo actúan las penalizaciones Ridge y Lasso, y qué efecto esperarías si las variables horas_entrenamiento y vo2max están muy correlacionadas.

**1.Ridge:** Le pondría un frenado suavce a los coeficientes para que no se vuelvan gigantescos. No apaga ñlas variablñes solo las encoge y repartiria el peso cuando hay variables muy iguales entre si. Si _horas entrenamiento_ y _vo2max_ estuvieran muy correlacionadas, las mantendría con pesos mas pequeños o estables.

**2.Lasso:** Como es mas agresivo, ademas de encoger, podría apagar las variabñes dejandolas en 0. Eso simplificaria el modelo y nos dejaría con lo más importante, eligiria una de las dos y descartaria la otra, lo que nos haria un modelo más compacto pero maybe de una forma arbitraria-

### Pregunta 4: Menciona una situación en la que preferirías usar un modelo no lineal (como un Random Forest o SVR) sobre una regresión lineal en este contexto.

Cuando la relación no sea una recta clara, por ejemplo que el tiempo mejorara rapido al subior el vo2max, tambien cuando hubiera interacciones fuertes entre variables, ejemplo si las horas de entrenamiento cambiaran segun lass lesiones previas o el sexo. También si los errores aumentaran o hubiera outliers. En estos casos usaria RF o SVR que capturarian mejor las curvas.



## Sección Práctica

In [10]:
# Pregunta 5: Carga el dataset y separa los datos en entrenamiento (80%) y prueba (20%).

import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV, KFold, cross_val_score
from sklearn.ensemble import RandomForestRegressor

# 1 Separamos x/y
assert "rendimiento" in df.columns, "No encuentro la columna 'rendimiento'."
X = df.drop(columns=["rendimiento"])
y = df["rendimiento"]

# 2 tipos de columna
num_cols = X.select_dtypes(include=["number"]).columns.tolist()
cat_cols = X.select_dtypes(include=["object", "category"]).columns.tolist()

# 3 imputación simple + OHE para cat + escala numérica
num_pipe = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

cat_pipe = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("ohe", OneHotEncoder(handle_unknown="ignore", drop=None))
])

pre = ColumnTransformer(transformers=[
    ("num", num_pipe, num_cols),
    ("cat", cat_pipe, cat_cols)
])

# 4 split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.20, random_state=42
)

# 5 pipeline con RL
pipe_lr = Pipeline(steps=[
    ("pre", pre),
    ("model", LinearRegression())
])

pipe_lr.fit(X_train, y_train)

# RMSE
mse_train = mean_squared_error(y_train, pred_train)  
mse_test  = mean_squared_error(y_test,  pred_test)   

rmse_train = np.sqrt(mse_train)
rmse_test  = np.sqrt(mse_test)

print(f"RMSE (train): {rmse_train:.3f}")
print(f"RMSE (test) : {rmse_test:.3f}")

# 7 overfitting?
gap = rmse_test - rmse_train
rel_gap = gap / rmse_test if rmse_test != 0 else np.nan
msg = "Posible overfitting" if (gap > 0 and rel_gap > 0.15) else "No overfitting"
print(f"Diagnóstico: {msg} (gap absoluto: {gap:.3f}, gap relativo: {rel_gap:.2%})")

RMSE (train): 1.868
RMSE (test) : 1.643
Diagnóstico: No overfitting (gap absoluto: -0.226, gap relativo: -13.74%)


In [11]:
# Pregunta 6: Ajusta un modelo con Ridge, buscando el hiperparámetro óptimo de regularización con GridSearchCV.

# reutilizamos X, y, X_train, X_test, y_train, y_test definidos antes
num_cols = X.select_dtypes(include=["number"]).columns.tolist()
cat_cols = X.select_dtypes(include=["object", "category"]).columns.tolist()

num_pipe = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

cat_pipe = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("ohe", OneHotEncoder(handle_unknown="ignore"))
])

pre = ColumnTransformer(transformers=[
    ("num", num_pipe, num_cols),
    ("cat", cat_pipe, cat_cols)
])

pipe_ridge = Pipeline(steps=[
    ("pre", pre),
    ("model", Ridge(random_state=42))
])

# buscamos alpha
param_grid = {
    "model__alpha": np.logspace(-3, 3, 13)  
}

cv = KFold(n_splits=5, shuffle=True, random_state=42)

gs = GridSearchCV(
    estimator=pipe_ridge,
    param_grid=param_grid,
    scoring="neg_mean_squared_error",
    cv=cv,
    n_jobs=-1,
    refit=True,
    verbose=0
)

gs.fit(X_train, y_train)

best_alpha = gs.best_params_["model__alpha"]
best_cv_mse = -gs.best_score_
best_cv_rmse = np.sqrt(best_cv_mse)

print(f"Mejor alpha (Ridge): {best_alpha}")
print(f"CV RMSE (5-fold) con mejor alpha: {best_cv_rmse:.3f}")

# evaluacion final con el mejor modelo
pred_train_ridge = gs.predict(X_train)
pred_test_ridge  = gs.predict(X_test)

rmse_train_ridge = np.sqrt(mean_squared_error(y_train, pred_train_ridge))
rmse_test_ridge  = np.sqrt(mean_squared_error(y_test,  pred_test_ridge))

print(f"RMSE Ridge (train): {rmse_train_ridge:.3f}")
print(f"RMSE Ridge (test) : {rmse_test_ridge:.3f}")

# comprobacion recalculanmdo base lineal
from sklearn.linear_model import LinearRegression

pipe_lr = Pipeline(steps=[
    ("pre", pre),
    ("model", LinearRegression())
])
pipe_lr.fit(X_train, y_train)
rmse_train_lr = np.sqrt(mean_squared_error(y_train, pipe_lr.predict(X_train)))
rmse_test_lr  = np.sqrt(mean_squared_error(y_test,  pipe_lr.predict(X_test)))

print(f"[Base] RMSE Lineal (train): {rmse_train_lr:.3f}")
print(f"[Base] RMSE Lineal (test) : {rmse_test_lr:.3f}")

mejor = "Ridge" if rmse_test_ridge < rmse_test_lr else "Lineal"
print(f"→ Prefiero: {mejor} (menor RMSE en test)")

Mejor alpha (Ridge): 10.0
CV RMSE (5-fold) con mejor alpha: 1.938
RMSE Ridge (train): 1.870
RMSE Ridge (test) : 1.645
[Base] RMSE Lineal (train): 1.868
[Base] RMSE Lineal (test) : 1.643
→ Prefiero: Lineal (menor RMSE en test)


In [12]:
# Pregunta 7: Ajusta un Random Forest Regressor.

# Reusamos X, y d
num_cols = X.select_dtypes(include=["number"]).columns.tolist()
cat_cols = X.select_dtypes(include=["object", "category"]).columns.tolist()

# preprocesamos
num_pipe = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median"))
])
cat_pipe = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("ohe", OneHotEncoder(handle_unknown="ignore"))
])

pre = ColumnTransformer(transformers=[
    ("num", num_pipe, num_cols),
    ("cat", cat_pipe, cat_cols)
])

rf = RandomForestRegressor(
    n_estimators=400,
    random_state=42,
    n_jobs=-1
)

pipe_rf = Pipeline(steps=[
    ("pre", pre),
    ("model", rf)
])

# validacion cruzada
cv = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores = cross_val_score(
    pipe_rf, X, y,
    scoring="neg_mean_squared_error",
    cv=cv,
    n_jobs=-1
)
cv_rmse = np.sqrt(-cv_scores)

print(f"CV RMSE (k=5) - promedio: {cv_rmse.mean():.3f} ± {cv_rmse.std():.3f}")

# train
pipe_rf.fit(X, y)

feat_names = pipe_rf.named_steps["pre"].get_feature_names_out()
importances = pipe_rf.named_steps["model"].feature_importances_

imp_series = pd.Series(importances, index=feat_names).sort_values(ascending=False)
top3 = imp_series.head(3)
print("\nTop 3 variables más importantes:")
for name, val in top3.items():
    print(f"- {name}: {val:.3f}")


CV RMSE (k=5) - promedio: 1.971 ± 0.186

Top 3 variables más importantes:
- num__vo2max: 0.318
- num__porcentaje_grasa: 0.277
- num__horas_entrenamiento: 0.182


## Sección de Reflexión

### Pregunta 8:  Supón que te piden implementar este modelo en un centro de alto rendimiento con recursos limitados.

1. Me quedo con Regresión Lineal. REindió mejor en test (RMSE 1.643) que Ridge 1.645 y es simple, barato de operar y fácil de explicar con recursos limitados. Si quisiera un extra de estabilidad frente a multicolinealidad, usaría Ridge con el α que ya encontramos, pero la base lineal cunmlio bien.

2. Mantendrí las variables clave: vo2max, porcentaje_grasa, horas_entrenamiento, que salieron en top de random forest y bhacen más lógica. Borraria suplementos que no fue relevante. Lo demás si se considero.

3. A un entrenador le diría que un vo2max esta dando tiempo smás bajos, y un % de grasa más bajo aumenta el rendimiento. lAs horas de entrenamiento ayuidan pero los resultados dependen mucho de otras variablñes del atleta. Considerando estos tres podriamos definir con mayo precision las horas de entrenamiento que realmente ofrecerían una mejora constante en el resto de variablñes, mejorando rendmimiento y triempos.